tests/integration-tests/tests/update/test_update_rollback_fixes.py

```diff
@@ -0,0 +1,496 @@
+
+    if not cfn_bin_path:
+        # Fallback: find the path
+        result = remote_command_executor.run_remote_command("find /opt -name cfn-signal -type f 2>/dev/null | head -1")
+        cfn_signal_path = result.stdout.strip()
+        if cfn_signal_path:
+            cfn_bin_path = cfn_signal_path.rsplit("/cfn-signal", 1)[0]
+        else:
+            # Default path
+            cfn_bin_path = "/opt/parallelcluster/pyenv/versions/3.12.11/envs/cfn_bootstrap_virtualenv/bin"
+
+    logger.info(f"CFN bin path: {cfn_bin_path}")
+
+    # Create a wrapper script that makes cfn-signal fail
+    # This simulates an expired wait condition handle
+    wrapper_script = """#!/bin/bash
+# Wrapper script to simulate cfn-signal failure (expired wait condition)
+# Log the call for debugging
+echo "$(date): cfn-signal called with args: $@" >> /tmp/cfn-signal-wrapper.log
+echo "Simulated cfn-signal failure: wait condition expired" >&2
+exit 1
+"""
+    # Write wrapper script
+    remote_command_executor.run_remote_command(
+        f"cat << 'EOF' | sudo tee /tmp/cfn-signal-wrapper.sh > /dev/null\n{wrapper_script}EOF"
+    )
+    remote_command_executor.run_remote_command("sudo chmod +x /tmp/cfn-signal-wrapper.sh")
+
+    # Backup original cfn-signal and replace with wrapper
+    # Note: CFN_BOOTSTRAP_VIRTUALENV_PATH already points to bin dir, so just append /cfn-signal
+    remote_command_executor.run_remote_command(f"sudo cp {cfn_bin_path}/cfn-signal {cfn_bin_path}/cfn-signal.bak")
+    remote_command_executor.run_remote_command(f"sudo cp /tmp/cfn-signal-wrapper.sh {cfn_bin_path}/cfn-signal")
+    logger.info("cfn-signal wrapper installed")
+
+
+def _disable_cfn_hup_on_compute_node(remote_command_executor, node_name):
+    """
+    Disable cfn-hup on a compute node using srun.
+
+    Uses supervisorctl to stop cfn-hup service on the compute node.
+    """
+    logger.info(f"Disabling cfn-hup on compute node {node_name}...")
+
+    # Get supervisorctl path
+    result = remote_command_executor.run_remote_command(
+        "find /opt/parallelcluster -name supervisorctl -type f 2>/dev/null | head -1"
+    )
+    supervisorctl_path = result.stdout.strip()
+    if not supervisorctl_path:
+        supervisorctl_path = "/opt/parallelcluster/pyenv/versions/3.12.11/envs/cookbook_virtualenv/bin/supervisorctl"
+
+    # Stop cfn-hup using srun
+    remote_command_executor.run_remote_command(f"srun -w {node_name} sudo {supervisorctl_path} stop cfn-hup")
+
+    # Verify cfn-hup is stopped
+    # Note: supervisorctl status returns exit code 3 when process is STOPPED, so we use raise_on_error=False
+    result = remote_command_executor.run_remote_command(
+        f"srun -w {node_name} sudo {supervisorctl_path} status cfn-hup",
+        raise_on_error=False,
+    )
+    assert_that(result.stdout).contains("STOPPED")
+    logger.info(f"cfn-hup stopped on {node_name} ✓")
+
+
+def _wait_for_stack_rollback_complete(cluster, region, timeout_minutes=60):
```

Comment:
Why do we need to nest functions here? If the only reason is to inject the timeout_minutes parameter, I suggest to simplify by removing the nested functions and use a static timeout that we think it is reasonable.